pkg/services/ngalert/ngalert.go

```diff
@@ -438,3 +441,48 @@ func configureHistorianBackend(ctx context.Context, cfg setting.UnifiedAlertingS
 
 	return nil, fmt.Errorf("unrecognized state history backend: %s", backend)
 }
+
+// applyStateHistoryFeatureToggles edits state history configuration to comply with currently active feature toggles.
+func applyStateHistoryFeatureToggles(cfg *setting.UnifiedAlertingStateHistorySettings, ft featuremgmt.FeatureToggles, logger log.Logger) {
+	backend, _ := historian.ParseBackendType(cfg.Backend)
+	// These feature toggles represent specific, common backend configurations.
+	// If all toggles are enabled, we listen to the state history config as written.
+	// If any of them are disabled, we ignore the configured backend and treat the toggles as an override.
+	// If multiple toggles are disabled, we go with the most "restrictive" one.
+	if !ft.IsEnabled("alertStateHistoryLokiSecondary") {
+		// If we cannot even treat Loki as a secondary, we must use annotations only.
+		if backend == historian.BackendTypeMultiple || backend == historian.BackendTypeLoki {
+			logger.Info("Forcing Annotation backend due to state history feature toggles")
+			cfg.Backend = "annotations"
+			cfg.MultiPrimary = ""
+			cfg.MultiSecondaries = make([]string, 0)
+		}
+		return
+	}
+	if !ft.IsEnabled("alertStateHistoryLokiPrimary") {
+		// If we're using multiple backends, Loki must be the secondary.
+		if backend == historian.BackendTypeMultiple {
+			logger.Info("Coercing Loki to a secondary backend due to state history feature toggles")
+			cfg.MultiPrimary = "annotations"
+			cfg.MultiSecondaries = []string{"loki"}
+		}
+		// If we're using loki, we are only allowed to use it as a secondary. Dual write to it, plus annotations.
+		if backend == historian.BackendTypeLoki {
+			logger.Info("Coercing Loki to dual writes with a secondary backend due to state history feature toggles")
+			cfg.Backend = "multiple"
+			cfg.MultiPrimary = "annotations"
+			cfg.MultiSecondaries = []string{"loki"}
```

Comment:
The only difference between this `if` condition and the previous one is the log message. The effect is the same:
- Backend type is "multiple".
- Primary backend is "annotations".
- Secondaries is a slice containing "loki".

Maybe we can de-duplicate this code. Also, there's no need to check both conditions, they are mutually exclusive.

Reply:
This change is not strictly necessary, you can leave it as it is for readability reasons.